rls/src/main/java/io/grpc/rls/internal/AdaptiveThrottler.java

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.rls.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/**
+ * Implementation of {@link Throttler} that keeps track of recent history (the duration of which is
+ * specified to the constructor) and throttles requests at the client side based on the number of
+ * requests that the
+ * backend has accepted and the total number of requests generated. A given request will be
+ * throttled with a probability
+ * <pre>
+ *   throttleProbability = (requests - ratio_for_accepts * accepts) / (requests + requests_padding)
+ * </pre>
+ * where requests is the total number of requests, accepts is the total number of requests that the
+ * backend has accepted and ratio_for_accepts is just a constant multiplier passed to the
+ * constructor (see the description of ratio_for_accepts for more information).
+ */
+public final class AdaptiveThrottler implements Throttler {
+
+  private static final int DEFAULT_HISTORY_SECONDS = 30;
+  private static final int DEFAULT_REQUEST_PADDING = 8;
+  private static final float DEFAULT_RATIO_FOR_ACCEPT = 1.2f;
+
+  /**
+   * The duration of history of calls used by Adaptive Throttler.
+   */
+  private final int historySeconds;
+  /**
+   * A magic number to tune the aggressiveness of the throttling. High numbers throttle less. The
+   * default is 8.
+   */
+  private final int requestsPadding;
+  /**
+   * The ratio by which the Adaptive Throttler will attempt to send requests above what the server
+   * is currently accepting.
+   */
+  private final float ratioForAccepts;
+  private final Ticker ticker;
+  private final Random random;
```

Comment:
Random is not thread safe. You can refactor `io.grpc.xds.ThreadSafeRandom` to intenal or util.

Reply:
i just used the `ThreadLocalRandom` because the random doesn't need to be passed from builder.